pages/glossary/visible-changes-in-content.md

```diff
@@ -0,0 +1,22 @@
+---
+title: Visible changes in content
+key: visible-changes-in-content
+unambiguous: true
+objective: true
+input_aspects:
+  - CSS styling
+  - DOM tree
+---
+
+A _[event][] originated change in the content_ of a [web page][] occurs when, by comparing the [web page][] before and after the event [firing][] the rendered pixels change in any part of the document that is currently within the [viewport][] or that can be brought into the [viewport][] via scrolling.
```

Comment:
In extension of https://github.com/act-rules/act-rules.github.io/pull/1396/files#r471423985, I'd argue that the wording "before and after" isn't objective. As you mention the rule being optimised for manual implementations, "before and after" will be different from the perspective of a human observer than it will from a software observer. Whereas the software observer will be able to determine visual changes caused by the firing of an event on the frame render immediately following the event, a human observer will only be able to determine visual changes caused by the firing of an event after _several_ frame renders. The definition of course seems perfectly unambiguous; both of these observations happen _after_ the firing of the event. However, they're relative to the observer and therefore not objective.

Reply:
@kasperisager I've been trying to find a definition for "frame renders" and I haven't been able. 

Another possibility I came up is to say there is a change if the event originates at least one reflow or repaint of the document. Problems:
- I also can't find a definition of reflow or repaint in the specs because this is terminology for how the user agents render a web page
- I'm not sure how an evaluator would be able to establish a direct relationship between the event and the reflow of repaint happening after the event was triggered 

Any more thoughts on this?

Reply:
I think that'd be the right direction to take. You might be able to use the concept of "animation frame callback" for this: https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#animation-frames. I'm thinking something along the lines of "on or after the next run of animation frame callbacks".

Reply:
I was also looking at that, but I'm not convinced that all reflows or repaints imply an "animation frame callback". The definition begins with "Some objects include the AnimationFrameProvider interface mixin." I couldn't find anywhere specifying which object are those. I'll try to dig a bit deeper.

Reply:
The processing model for the event loop does seem to imply that animation frame callbacks are run whenever the rendering is updated: https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model:run-the-animation-frame-callbacks-3